R/assert.R

```diff
@@ -1,31 +1,59 @@
 #' Assert an argument has known prototype and/or size
 #'
-#' If the prototype doesn't match, an error of class
-#' `"vctrs_error_assert_ptype"` is raised.
-#' If the prototype doesn't match, an error of class
+#' @description
+#'
+#' * `vec_is()` is a predicate that checks if its input conforms to a
+#'   prototype and/or a size.
+#'
+#' * `vec_assert()` throws an error when the input doesn't conform.
+#'
+#' @section Error types:
+#'
+#' * If the prototype doesn't match, an error of class
+#'   `"vctrs_error_assert_ptype"` is raised.
+#'
+#' * If the prototype doesn't match, an error of class
 #' `"vctrs_error_assert_size"` is raised.
+#'
 #' Both errors inherit from `"vctrs_error_assert"`.
 #'
 #' @param x A vector argument to check.
 #' @param ptype Prototype to compare against.
 #' @param size Size to compare against
-#' @return Either an error of class `"vctrs_error_assert"`, or `x`, invisibly.
+#' @param arg Name of argument being checked. This is used in error
+#'   messages. The label of the expression passed as `x` is taken as
+#'   default.
+#'
+#' @return `vec_is()` returns `TRUE` or `FALSE`. `vec_assert()` either
+#'   throws a typed error (see section on error types) or returns `x`,
+#'   invisibly.
+#' @export
+vec_assert <- function(x, ptype = NULL, size = NULL, arg = NULL) {
+  arg <- arg %||% as_label(substitute(x))
+  vec_is_impl(x, arg, ptype, size, assert = TRUE)
+}
 #' @export
-vec_assert <- function(x, ptype = NULL, size = NULL) {
-  x_name <- as_label(substitute(x))
+vec_is <- function(x, ptype = NULL, size = NULL) {
+  vec_is_impl(x, "", ptype, size)
+}
 
+vec_is_impl <- function(x, arg, ptype = NULL, size = NULL, assert = FALSE) {
   if (!is.null(ptype)) {
     x_type <- vec_type(x)
     ptype <- vec_type(ptype)
 
     if (!identical(ptype, x_type)) {
-      msg <- paste0("`", x_name, "` must be <", vec_ptype_abbr(ptype), ">, not <", vec_ptype_abbr(x_type), ">.")
-      abort(
-        msg,
-        .subclass = c("vctrs_error_assert_ptype", "vctrs_error_assert"),
-        required = ptype,
-        actual = x_type
-      )
+      if (assert) {
```

Comment:
Maybe refactor out `has_size()` and `has_type()` functions so that `vec_is()` can just be `has_size(x, size) && has_type(x,  ptype)` and `vec_assert()` can focus on throwing errors?

Reply:
This would require `vec_assert()` to dispatch twice for type and twice for size (when it becomes generic), making it more expensive than really needed. This might discourage users from using vector assertions.

Reply:
Given that you'll generally check once at the start of a function, I don't think performance should be a dominant concern here.

Reply:
Won't functions created with vctrs be used in loops, e.g. in a grouped mutate? I thought the whole package needed to be performant.

Reply:
This just feels like over eager performance optimisation — I think creating a type-unstable function like this (even though it's internal) dominates potential future concerns about performance.

Reply:
Done: https://github.com/r-lib/vctrs/pull/226/commits/05dd79df7274173fe7df4bd925bf240e122e4be4

This approach also suffers from code duplication. I think the original approach was indeed type unstable, but in a structured way. Just like gotos can be used in a structured, predictable, and readable way.

Reply:
hmm I guess with this duplication it doesn't make sense to call `has_ptype()` and dispatch twice, we should just inline it. Or am I missing an obvious way? Are we overthinking this little bit of code? ;)

Reply:
Does this make more sense? https://github.com/r-lib/vctrs/pull/226/commits/810aed54de492aa25b4b770425caf32ce47899d2